File: 376-wiggle-subsequence/376-wiggle-subsequence.cpp
class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        if (nums.size() == 1) {
            return 1;
        }
        if (nums.size() == 2) {
            if (nums[0] != nums[1]) {
                return 2;
            }
            return 1;
        }
        int count = 1;
        int increasing = 2;  // 0 false, 1 true, 2 None
        for (int i=1; i<nums.size(); i++) {
            if (nums[i] != nums[i-1]) {
                if (increasing == 2) {
                    if (nums[i] < nums[i-1]) {
                        increasing = 0;
                    }
                    else {
                        increasing = 1;
                    }
                }
                if (increasing == 1) {
                    if (nums[i] > nums[i-1]) {
                        increasing = 0;
                        count++;
                    }
                }
                else {
                    if (nums[i] < nums[i-1]) {
                        increasing = 1;
                        count++;
                    }
                }
            }
        }
        return count;
    }
};

File: 376-wiggle-subsequence/376-wiggle-subsequence.py
class Solution:
    def wiggleMaxLength(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return 1
        if len(nums) == 2:
            if nums[0] != nums[1]:
                return 2
            return 1
        count = 1
        increasing = None
        for i in range(1, len(nums)):
            if nums[i] != nums[i-1]:
                if increasing is None:
                    if nums[i] < nums[i-1]:
                        increasing = False
                    else:
                        increasing = True
                if increasing:
                    if nums[i] > nums[i-1]:
                        increasing = False
                        count += 1
                else:
                    if nums[i] < nums[i-1]:
                        increasing = True
                        count += 1
        return count
